File: MCTS_BlokusDuo_cpp/GamePiece.cpp
//
// Created by joaovieira on 12/24/19.
//

#include "GamePiece.h"

const int GamePiece::transformations[8][2][2] = {
    {{ 1,  0}, { 0,  1}},
    {{-1,  0}, { 0,  1}},
    {{ 0, -1}, { 1,  0}},
    {{ 0,  1}, { 1,  0}},
    {{-1,  0}, { 0, -1}},
    {{ 1,  0}, { 0, -1}},
    {{ 0,  1}, {-1,  0}},
    {{ 0, -1}, {-1,  0}}
};

// TODO optimize array allocation
GamePiece::GamePiece(char codeName, std::vector<std::vector<int>> squares, std::vector<std::vector<int>> anchors) {
    this->codeName = codeName;
    this->transformation = 0;

    this->squares = new std::vector<Position *>();
    for (auto & square : squares)
        this->squares->push_back(new Position(square[0], square[1]));

    this->anchors = new std::vector<Position *>();
    for (auto & anchor : anchors)
        this->anchors->push_back(new Position(anchor[0], anchor[1]));
}

GamePiece::~GamePiece() {
    for (auto & anchor : *anchors)
        delete anchor;
    delete anchors;

    for (auto & square : *squares)
        delete square;
    delete squares;
}

std::list<std::vector<GamePiece *> *> *GamePiece::getInitialSetOfGamePieces() {
    auto initialSetOfGamePieces = new std::list<std::vector<GamePiece *> *>();

    struct DefaultGamePiece {char codeName; std::vector<std::vector<int>> squares, anchors; std::vector<int> variants;};
    std::vector<struct DefaultGamePiece> defaultGamePieces = {
            {.codeName = 'a', .squares = {{ 0,  0}},                                         .anchors = {{ 0,  0}},                                         .variants = {0}},
            {.codeName = 'b', .squares = {{ 0,  0}, { 0,  1}},                               .anchors = {{ 0,  0}, { 0,  1}},                               .variants = {0, 2}},
            {.codeName = 'c', .squares = {{ 0,  0}, { 0,  1}, { 0, -1}},                     .anchors = {{ 0,  1}, { 0, -1}},                               .variants = {0, 2}},
            {.codeName = 'd', .squares = {{ 0,  0}, { 0, -1}, { 1,  0}},                     .anchors = {{ 0,  0}, { 0, -1}, { 1,  0}},                     .variants = {0, 1, 2, 3}},
            {.codeName = 'e', .squares = {{ 0,  0}, { 0,  1}, { 0,  2}, { 0, -1}},           .anchors = {{ 0,  2}, { 0, -1}},                               .variants = {0, 2}},
            {.codeName = 'f', .squares = {{ 0,  0}, { 0,  1}, {-1,  1}, { 0, -1}},           .anchors = {{ 0,  1}, {-1,  1}, { 0, -1}},                     .variants = {0, 1, 2, 3, 4, 5, 6, 7}},
            {.codeName = 'g', .squares = {{ 0,  0}, { 0,  1}, { 0, -1}, { 1,  0}},           .anchors = {{ 0,  1}, { 1,  0}, { 0, -1}},                     .variants = {0, 1, 2, 6}},
            {.codeName = 'h', .squares = {{ 0,  0}, { 1,  0}, { 0,  1}, { 1,  1}},           .anchors = {{ 0,  0}, { 1,  0}, { 0,  1}, { 1,  1}},           .variants = {0}},
            {.codeName = 'i', .squares = {{ 0,  0}, {-1,  0}, { 0,  1}, { 1,  1}},           .anchors = {{ 0,  0}, {-1,  0}, { 0,  1}, { 1,  1}},           .variants = {0, 1, 2, 3}},
            {.codeName = 'j', .squares = {{ 0,  0}, { 0, -1}, { 0, -2}, { 0,  1}, { 0,  2}}, .anchors = {{ 0, -2}, { 0,  2}},                               .variants = {0, 2}},
            {.codeName = 'k', .squares = {{ 0,  0}, { 0, -1}, { 0, -2}, { 0,  1}, {-1,  1}}, .anchors = {{ 0, -2}, { 0,  1}, {-1,  1}},                     .variants = {0, 1, 2, 3, 4, 5, 6, 7}},
            {.codeName = 'l', .squares = {{ 0,  0}, { 0, -1}, { 0, -2}, {-1,  0}, {-1,  1}}, .anchors = {{ 0, -2}, {-1,  0}, {-1,  1}},                     .variants = {0, 1, 2, 3, 4, 5, 6, 7}},
            {.codeName = 'm', .squares = {{ 0,  0}, { 0, -1}, { 0,  1}, {-1,  0}, {-1,  1}}, .anchors = {{ 0, -1}, {-1,  0}, {-1,  1}, { 0,  1}},           .variants = {0, 1, 2, 3, 4, 5, 6, 7}},
            {.codeName = 'n', .squares = {{ 0,  0}, { 0, -1}, { 0,  1}, {-1, -1}, {-1,  1}}, .anchors = {{-1, -1}, { 0, -1}, {-1,  1}, { 0,  1}},           .variants = {0, 1, 2, 6}},
            {.codeName = 'o', .squares = {{ 0,  0}, { 0, -1}, { 1,  0}, { 0,  1}, { 0,  2}}, .anchors = {{ 0,  2}, { 1,  0}, { 0, -1}},                     .variants = {0, 1, 2, 3, 4, 5, 6, 7}},
            {.codeName = 'p', .squares = {{ 0,  0}, { 0, -1}, { 0,  1}, {-1,  1}, { 1,  1}}, .anchors = {{ 0, -1}, {-1,  1}, { 1,  1}},                     .variants = {0, 2, 3, 4}},
            {.codeName = 'q', .squares = {{ 0,  0}, { 0, -1}, { 0, -2}, { 1,  0}, { 2,  0}}, .anchors = {{ 0,  0}, { 0, -2}, { 2,  0}},                     .variants = {0, 1, 2, 3}},
            {.codeName = 'r', .squares = {{ 0,  0}, { 0, -1}, {-1, -1}, { 1,  0}, { 1,  1}}, .anchors = {{ 0,  0}, { 0, -1}, {-1, -1}, { 1,  0}, { 1,  1}}, .variants = {0, 1, 2, 3}},
            {.codeName = 's', .squares = {{ 0,  0}, { 1,  0}, { 1,  1}, {-1,  0}, {-1, -1}}, .anchors = {{ 1,  0}, { 1,  1}, {-1,  0}, {-1, -1}},           .variants = {0, 1, 2, 3}},
            {.codeName = 't', .squares = {{ 0,  0}, { 1,  0}, { 0,  1}, {-1,  0}, {-1, -1}}, .anchors = {{ 1,  0}, { 0,  1}, {-1,  0}, {-1, -1}},           .variants = {0, 1, 2, 3, 4, 5, 6, 7}},
            {.codeName = 'u', .squares = {{ 0,  0}, { 1,  0}, { 0,  1}, {-1,  0}, { 0, -1}}, .anchors = {{ 1,  0}, { 0,  1}, {-1,  0}, { 0, -1}},           .variants = {0}}
    };

    for (const auto& defaultGamePiece : defaultGamePieces) {
        auto gamePiece = new std::vector<GamePiece *>();

        for (const auto& transformation : defaultGamePiece.variants) {
            auto variant = new GamePiece(defaultGamePiece.codeName, defaultGamePiece.squares, defaultGamePiece.anchors);
            variant->transformPiece(transformation);
            gamePiece->push_back(variant);
        }

        initialSetOfGamePieces->push_back(gamePiece);
    }

    return initialSetOfGamePieces;
}

char GamePiece::getCodeName() {
    return codeName;
}

void GamePiece::transformPiece(int transformation) {
    this->transformation = transformation;

    for (auto & square : *squares)
        square->transform(transformations[transformation]);

    for (auto & anchor : *anchors)
        anchor->transform(transformations[transformation]);
}

std::vector<Position *> *GamePiece::getSquares() {
    return squares;
}

std::vector<Position *> *GamePiece::getAnchors() {
    return anchors;
}

int GamePiece::getTransformation() {
    return transformation;
}

bool GamePiece::equals(GamePiece *gamePiece) {
    if (this == gamePiece) return true;
    if (gamePiece == nullptr) return false;
    return codeName == gamePiece->codeName && transformation == gamePiece->transformation;
}

File: MCTS_BlokusDuo_java/src/com/joaomiguelvieira/GamePiece.java
package com.joaomiguelvieira;

import java.lang.reflect.Array;
import java.util.*;

public class GamePiece {
    private static final int DEFAULT_MAX_SQUARES = 5;
    private static final int[][][] transformations = {
            {{ 1,  0}, { 0,  1}},
            {{-1,  0}, { 0,  1}},
            {{ 0, -1}, { 1,  0}},
            {{ 0,  1}, { 1,  0}},
            {{-1,  0}, { 0, -1}},
            {{ 1,  0}, { 0, -1}},
            {{ 0,  1}, {-1,  0}},
            {{ 0, -1}, {-1,  0}}
    };

    private char codeName;
    private int transformation;
    private Position[] squares, anchors;

    public GamePiece(char codeName, int[][] squares, int[][] anchors) {
        this.codeName = codeName;
        this.transformation = 0;

        this.squares = new Position[squares.length];
        for (int i = 0; i < squares.length; i++)
            this.squares[i] = new Position(squares[i][0], squares[i][1]);

        this.anchors = new Position[anchors.length];
        for (int i = 0; i < anchors.length; i++)
            this.anchors[i] = new Position(anchors[i][0], anchors[i][1]);
    }

    public static ArrayList<ArrayList<GamePiece>> getInitialSetOfGamePieces () {
        ArrayList<ArrayList<GamePiece>> initialSetOfGamePieces = new ArrayList<>();

        class DefaultGamePiece {
            public char codeName;
            public int[][] squares;
            public int[][] anchors;
            public int[] variants;

            DefaultGamePiece(char codeName, int[][] squares, int[][] anchors, int[] variants) {
                this.codeName = codeName;
                this.squares = squares;
                this.anchors = anchors;
                this.variants = variants;
            }
        }

        DefaultGamePiece[] defaultGamePieces = {
                new DefaultGamePiece('a', new int[][]{{ 0,  0}},                                         new int[][]{{ 0,  0}}                                        , new int[]{0}),
                new DefaultGamePiece('b', new int[][]{{ 0,  0}, { 0,  1}},                               new int[][]{{ 0,  0}, { 0,  1}}                              , new int[]{0, 2}),
                new DefaultGamePiece('c', new int[][]{{ 0,  0}, { 0,  1}, { 0, -1}},                     new int[][]{{ 0,  1}, { 0, -1}, { 0,  0}}                    , new int[]{0, 2}),
                new DefaultGamePiece('d', new int[][]{{ 0,  0}, { 0, -1}, { 1,  0}},                     new int[][]{{ 0,  0}, { 0, -1}, { 1,  0}}                    , new int[]{0, 1, 2, 3}),
                new DefaultGamePiece('e', new int[][]{{ 0,  0}, { 0,  1}, { 0,  2}, { 0, -1}},           new int[][]{{ 0,  2}, { 0, -1}, { 0,  0}, { 0,  0}}          , new int[]{0, 2}),
                new DefaultGamePiece('f', new int[][]{{ 0,  0}, { 0,  1}, {-1,  1}, { 0, -1}},           new int[][]{{ 0,  1}, {-1,  1}, { 0, -1}, { 0,  0}}          , new int[]{0, 1, 2, 3, 4, 5, 6, 7}),
                new DefaultGamePiece('g', new int[][]{{ 0,  0}, { 0,  1}, { 0, -1}, { 1,  0}},           new int[][]{{ 0,  1}, { 1,  0}, { 0, -1}, { 0,  0}}          , new int[]{0, 1, 2, 6}),
                new DefaultGamePiece('h', new int[][]{{ 0,  0}, { 1,  0}, { 0,  1}, { 1,  1}},           new int[][]{{ 0,  0}, { 1,  0}, { 0,  1}, { 1,  1}}          , new int[]{0}),
                new DefaultGamePiece('i', new int[][]{{ 0,  0}, {-1,  0}, { 0,  1}, { 1,  1}},           new int[][]{{ 0,  0}, {-1,  0}, { 0,  1}, { 1,  1}}          , new int[]{0, 1, 2, 3}),
                new DefaultGamePiece('j', new int[][]{{ 0,  0}, { 0, -1}, { 0, -2}, { 0,  1}, { 0,  2}}, new int[][]{{ 0, -2}, { 0,  2}, { 0,  0}, { 0,  0}, { 0,  0}}, new int[]{0, 2}),
                new DefaultGamePiece('k', new int[][]{{ 0,  0}, { 0, -1}, { 0, -2}, { 0,  1}, {-1,  1}}, new int[][]{{ 0, -2}, { 0,  1}, {-1,  1}, { 0,  0}, { 0,  0}}, new int[]{0, 1, 2, 3, 4, 5, 6, 7}),
                new DefaultGamePiece('l', new int[][]{{ 0,  0}, { 0, -1}, { 0, -2}, {-1,  0}, {-1,  1}}, new int[][]{{ 0, -2}, {-1,  0}, {-1,  1}, { 0,  0}, { 0,  0}}, new int[]{0, 1, 2, 3, 4, 5, 6, 7}),
                new DefaultGamePiece('m', new int[][]{{ 0,  0}, { 0, -1}, { 0,  1}, {-1,  0}, {-1,  1}}, new int[][]{{ 0, -1}, {-1,  0}, {-1,  1}, { 0,  1}, { 0,  0}}, new int[]{0, 1, 2, 3, 4, 5, 6, 7}),
                new DefaultGamePiece('n', new int[][]{{ 0,  0}, { 0, -1}, { 0,  1}, {-1, -1}, {-1,  1}}, new int[][]{{-1, -1}, { 0, -1}, {-1,  1}, { 0,  1}, { 0,  0}}, new int[]{0, 1, 2, 6}),
                new DefaultGamePiece('o', new int[][]{{ 0,  0}, { 0, -1}, { 1,  0}, { 0,  1}, { 0,  2}}, new int[][]{{ 0,  2}, { 1,  0}, { 0, -1}, { 0,  0}, { 0,  0}}, new int[]{0, 1, 2, 3, 4, 5, 6, 7}),
                new DefaultGamePiece('p', new int[][]{{ 0,  0}, { 0, -1}, { 0,  1}, {-1,  1}, { 1,  1}}, new int[][]{{ 0, -1}, {-1,  1}, { 1,  1}, { 0,  0}, { 0,  0}}, new int[]{0, 2, 3, 4}),
                new DefaultGamePiece('q', new int[][]{{ 0,  0}, { 0, -1}, { 0, -2}, { 1,  0}, { 2,  0}}, new int[][]{{ 0,  0}, { 0, -2}, { 2,  0}, { 0,  0}, { 0,  0}}, new int[]{0, 1, 2, 3}),
                new DefaultGamePiece('r', new int[][]{{ 0,  0}, { 0, -1}, {-1, -1}, { 1,  0}, { 1,  1}}, new int[][]{{ 0,  0}, { 0, -1}, {-1, -1}, { 1,  0}, { 1,  1}}, new int[]{0, 1, 2, 3}),
                new DefaultGamePiece('s', new int[][]{{ 0,  0}, { 1,  0}, { 1,  1}, {-1,  0}, {-1, -1}}, new int[][]{{ 1,  0}, { 1,  1}, {-1,  0}, {-1, -1}, { 0,  0}}, new int[]{0, 1, 2, 3}),
                new DefaultGamePiece('t', new int[][]{{ 0,  0}, { 1,  0}, { 0,  1}, {-1,  0}, {-1, -1}}, new int[][]{{ 1,  0}, { 0,  1}, {-1,  0}, {-1, -1}, { 0,  0}}, new int[]{0, 1, 2, 3, 4, 5, 6, 7}),
                new DefaultGamePiece('u', new int[][]{{ 0,  0}, { 1,  0}, { 0,  1}, {-1,  0}, { 0, -1}}, new int[][]{{ 1,  0}, { 0,  1}, {-1,  0}, { 0, -1}, { 0,  0}}, new int[]{0})
        };

        for (DefaultGamePiece defaultGamePiece : defaultGamePieces) {
            ArrayList<GamePiece> newDefaultGamePiece = new ArrayList<>();

            for (int variant : defaultGamePiece.variants) {
                GamePiece newVariant = new GamePiece(defaultGamePiece.codeName, defaultGamePiece.squares, defaultGamePiece.anchors);
                newVariant.transformPiece(variant);
                newDefaultGamePiece.add(newVariant);
            }

            initialSetOfGamePieces.add(newDefaultGamePiece);
        }


        return initialSetOfGamePieces;
    }

    public char getCodeName() {
        return codeName;
    }

    public void transformPiece(int transformation) {
        this.transformation = transformation;

        for (Position square : squares)
            square.transform(transformations[transformation]);

        for (Position anchor : anchors)
            anchor.transform(transformations[transformation]);
    }

    public Position[] getSquares() {
        return squares;
    }

    public Position[] getAnchors() {
        return anchors;
    }

    public int getTransformation() {
        return transformation;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        GamePiece gamePiece = (GamePiece) o;
        return codeName == gamePiece.codeName &&
                transformation == gamePiece.transformation;
    }

    @Override
    public int hashCode() {
        return Objects.hash(codeName, transformation);
    }
}
